data_model/src/query/mod.rs

```diff
@@ -31,199 +23,207 @@ use self::{
-    events::TriggeringEventFilterBox,
-    seal,
-    transaction::{CommittedTransaction, SignedTransaction, TransactionPayload},
-    IdBox, Identifiable, IdentifiableBox,
+    domain::Domain,
+    parameter::{Parameter, Parameters},
+    peer::Peer,
+    permission::Permission,
+    role::{Role, RoleId},
+    seal::Sealed,
+    transaction::{CommittedTransaction, SignedTransaction},
+    trigger::TriggerId,
 };
 
-pub mod cursor;
-pub mod pagination;
+pub mod builder;
+pub mod parameters;
 pub mod predicate;
-pub mod sorting;
-
-/// Default value for `fetch_size` parameter in queries.
-pub const DEFAULT_FETCH_SIZE: NonZeroU32 = nonzero!(10_u32);
-
-/// Max value for `fetch_size` parameter in queries.
-pub const MAX_FETCH_SIZE: NonZeroU32 = nonzero!(10_000_u32);
-
-/// Structure for query fetch size parameter encoding/decoding
-#[derive(
-    Debug,
-    Default,
-    Clone,
-    Copy,
-    PartialEq,
-    Eq,
-    Constructor,
-    Decode,
-    Encode,
-    Deserialize,
-    Serialize,
-    IntoSchema,
-)]
-pub struct FetchSize {
-    /// Inner value of a fetch size.
-    ///
-    /// If not specified then [`DEFAULT_FETCH_SIZE`] is used.
-    pub fetch_size: Option<NonZeroU32>,
-}
-
-macro_rules! queries {
-    ($($($meta:meta)* $item:item)+) => {
-        pub use self::model::*;
-
-        #[iroha_data_model_derive::model]
-        mod model{
-            use super::*; $(
-
-            #[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
-            #[derive(parity_scale_codec::Decode, parity_scale_codec::Encode)]
-            #[derive(serde::Deserialize, serde::Serialize)]
-            #[derive(derive_more::Constructor)]
-            #[derive(iroha_schema::IntoSchema)]
-            $($meta)*
-            $item )+
-        }
-    };
-}
-
-/// Trait for typesafe query output
-pub trait Query: Into<QueryBox> + seal::Sealed {
-    /// Output type of query
-    type Output: Into<QueryOutputBox> + TryFrom<QueryOutputBox>;
 
-    /// [`Encode`] [`Self`] as [`QueryBox`].
-    ///
-    /// Used to avoid an unnecessary clone
-    fn encode_as_query_box(&self) -> Vec<u8>;
+/// A query that either returns a single value or errors out
+pub trait SingularQuery: Sealed {
+    /// The type of the output of the query
+    type Output;
 }
 
-/// A [`Query`] that either returns a single value or errors out
-pub trait SingularQuery: Query {}
-
-/// A [`Query`] that returns an iterable collection of values
-pub trait IterableQuery: Query {
-    /// A type of single element of the output collection
-    type Item;
+/// A query that returns an iterable collection of values
+///
+/// Iterable queries logically return a stream of items.
+/// In the actual implementation, the items collected into batches and a cursor is used to fetch the next batch.
+/// [`builder::IterableQueryIterator`] abstracts over this and allows the query consumer to use a familiar [`Iterator`] interface to iterate over the results.
+pub trait IterableQuery: Sealed {
```

Comment:
```suggestion
pub trait Query: Sealed {
```

in line with other changes I would make the fact that it's `Iterable` implicit and have `Singular` as a special case

Reply:
I am fine with making iterable queries the default case in the API methods, but would prefer to keep the trait names explicit. You would use them more on the implementation side, where I feel it's more important to think of the type of the query you are working with

Reply:
If we're going to remove `SingularQuery`, then I think you should rename it. But if you disagree, resolve the conversation

Reply:
I am not confident that singular queries are possible to be removed (#4933), so I am ready to commit to the rename.